### PR TITLE
Add functionality to modify app logo from plugin

### PR DESCRIPTION
**Pull request by ashu8912**
# [Title: describe the change in one sentence]
Setup headlamp to handle logo change via plugin

### CODE REVIEW

File: plugins/headlamp-plugin/package.json
Comment (illume) [inline plugins/headlamp-plugin/package.json:31]: Seems typescript was removed accidentally.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Branding is a term that encompasses several aspects about the visuals of a product.
So we should not really mention branding as the thing the plugin is changing if it's changing just the logo.

File: plugins/headlamp-plugin/package.json
Comment (ashu8912) [reply plugins/headlamp-plugin/package.json:42]: fixed

**Comment by ashu8912** [comment]
One part of #277

File: frontend/src/plugin/registry.tsx
Comment (joaquimrocha) [inline frontend/src/plugin/registry.tsx:140]: I thought I had left already a comment in a different review explaining that "branding" is a broad term, and that this plugin changes the app logo which should be specific name. Maybe `registerMainLogo`.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]

**Comment by ashu8912** [comment]
@joaquimrocha the CI is failing here I am not sure why it didn't take the patch about not using localeTime in stories

File: frontend/src/components/pod/__snapshots__/Pod.stories.storyshot
Comment (joaquimrocha) [inline frontend/src/components/pod/__snapshots__/Pod.stories.storyshot:216]: This is what's making the CI fail.

File: frontend/src/plugin/registry.tsx
Comment (joaquimrocha) [inline frontend/src/plugin/registry.tsx:28]: This looks a bit odd. Maybe we can make it be just be an object? (so it doesn't have to be used from Ts)

@illume , do you have a suggestion for this?

File: frontend/src/redux/reducers/ui.tsx
Comment (joaquimrocha) [inline frontend/src/redux/reducers/ui.tsx:201]: This is not a component, it's a function that returns a component.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Left a few comments, but it's almost there.

File: frontend/src/plugin/registry.tsx
Comment (illume) [reply frontend/src/plugin/registry.tsx:151]: What about `registerAppLogoComponent`? 
(This would require the function signature being changed.)

File: frontend/src/redux/reducers/ui.tsx
Comment (illume) [reply frontend/src/redux/reducers/ui.tsx:201]: I made a suggestion elsewhere that we pass in a component instead. This would need to be kept if we use that suggestion.

File: frontend/src/plugin/registry.tsx
Comment (illume) [reply frontend/src/plugin/registry.tsx:28]: It can be written a bit more clearly like this: `"small" | "large"`

If we pass in a component instead this would turn into the props interface. The function type here seems to already use props.

File: frontend/src/components/Sidebar/HeadlampButton.tsx
Comment (illume) [inline frontend/src/components/Sidebar/HeadlampButton.tsx:70]: You could get rid of LOGO_TYPE here if you used `"small" : "large"` and relied on typescript to check the strings.

File: frontend/src/plugin/registry.tsx
Comment (ashu8912) [reply frontend/src/plugin/registry.tsx:144]: @joaquimrocha I am a little confused are you suggesting to return a component or a function that returns a component? Using a component made more sense because All the state related info can be passed to the component via prop.
cc: @illume

**Comment by illume** [comment]
@ashu8912 I pushed a couple of small type check fixes causing the ci to fail `npm run tsc`.

File: frontend/src/plugin/registry.tsx
Comment (illume) [inline frontend/src/plugin/registry.tsx:143]: This example seems wrong?

File: frontend/src/plugin/registry.tsx
Comment (illume) [reply frontend/src/plugin/registry.tsx:146]: I made a suggestion here: https://github.com/kinvolk/headlamp/pull/429#issuecomment-1068571834 resolving this to clean things up.

**Comment by ashu8912** [comment]
Thanks @illume this is indeed a better approach.

Adding my comment here for why 
1. Storing directly the component into the store is a problem
2. Storing a function that returns a component into store is a problem

For the first case we would run into problems of inline component rendering where our store component gets rerendered each time the component that is asking it from the store is rerendered.

For the second case we would overcome the first problem because now our component reference is retained but this is still not recommended because functions are non-serializable and storing non serializable properties in redux store is not recommended.

**Comment by illume** [comment]
Looks like the fixes for the typescript errors were overwritten. See `npm run tsc`.

**Comment by joaquimrocha** [comment]
> @ashu8912 I pushed a couple of small type check fixes causing the ci to fail `npm run tsc`.
> Ok... most of how we could do it is here (untested and not very refined...):
> 
> ```tsx
> // Example How we use the API from the plugin
> 
> function MyLogo({logoType, theme}: {
>   logoType?: 'small' | 'large';
>   theme?: string;
>   [key: string]: any;
> }) {
>   return 
> }
> registerAppLogo(MyLogo)
> 
> 
> // Implementation of registerAppLogo
> registerAppLogo(Component: null | ReactElement) {

I am a bit confused. This Component arg is null or a ReactElement, but its being used above with a function, which AFAIU is not a ReactElement, it just does return one, but we're not calling the function.

>   setPluginAppLogo(Component)
>   // note we don't send the component to the redux action
>   store.dispatch(setAppLogo(component === null ? false : true));
> }
> 
> // Here we set the logo component singleton
> let pluginLogo = null; 
> function setPluginAppLogo(component : null | ReactElement) { 
>   pluginLogo = component
> }
> function getPluginAppLogo() : null | ReactElement {
>   return pluginLogo;
> }

This does accomplish what we want to do. If using functions as we've been doing in Redux is not recommendable, then maybe we can just have a global action store to abstract cases like the logo and others? The idea would be to have keys stored in redux, and using those to access this global object where the functions are defined.

We sort of had names for actions in the beginning in order to have a more controllable way of using such keys (though it was intended inside Redux). But of course, using unique keys (generated or logical, like "main-app-logo") may make sense instead.

> 
> // Now we select the state, and if there's a logo, we can get that logo.
> const pluginLogo = useTypedSelector(state => state.ui.branding.logo);
> const arePluginsLoaded = useTypedSelector(state => state.ui.pluginsLoaded);
> let PluginAppLogoComponent;

Right. With my suggestion, this could be a hook, like:
const getPluginAppLogo = usePluginComponent('name-of-my-key');
 
> if (pluginLogo) {
>   PluginAppLogoComponent = getPluginAppLogo();
> }
> ```

**Comment by illume** [comment]
Yeah, a usePluginComponent with keys would be a nice way to extend things via plugins.
```tsx
const getPluginAppLogo = usePluginComponent('name-of-my-key');
```

> I am a bit confused. This Component arg is null or a ReactElement, but its being used above with a function, which AFAIU is not a ReactElement, it just does return one, but we're not calling the function.

MyLogo there is a functional component, and it would take all the props a logo needs.

File: frontend/src/redux/actions/actions.tsx
Comment (joaquimrocha) [inline frontend/src/redux/actions/actions.tsx:30]: I still find this a bit breaking from the current way we've used, but we probably require a full review into how to unify and make sure the API is as future proof as possible.
If @illume is fine with this, let's let it through.

File: frontend/src/redux/actions/actions.tsx
Comment (joaquimrocha) [inline frontend/src/redux/actions/actions.tsx:23]: All the other action names here have verbs, so we should be consistent and call this `UI_BRANDING_SET_APP_LOGO`.

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
Still had a few comments going through this, but it's getting there.

File: frontend/src/components/Sidebar/HeadlampButton.tsx
Comment (ashu8912) [reply frontend/src/components/Sidebar/HeadlampButton.tsx]: oops, fixed

**Comment by illume** [comment]
I think the plugin API for changing the logo is fine 👍

Probably it would be nice going forward if we use new plugin APIs inside an example plugin - but I don't think that should block this being merged. For the implementation, I'd prefer it was done using a more generic extension point as discussed in the meeting `const AppLogo = usePluginComponent('logo', DefaultLogo);`. But I don't think this should block this being merged either.

**Review by joaquimrocha — APPROVED** [review]
LGTM.
Let's wait on getting #458 to be fixed before we merge this, so I don't have to fiddle with git before publishing a new headlamp-plugin.

**Comment by joaquimrocha** [comment]
> LGTM. Let's wait on getting #458 to be fixed before we merge this, so I don't have to fiddle with git before publishing a new headlamp-plugin.

Since the fix was in the backend, then there's no problem with merging this, as we won't have a new release for headlamp-plugin.

**Pull request merged**
